modules/post/windows/recon/resolve_ip.rb

```diff
@@ -27,35 +32,66 @@ def initialize(info = {})
         }
       )
     )
-    register_options(
-      [
-        OptAddress.new("ADDRESS", [ false, "Enumerate currently configured shares"]),
-        OptAddressRange.new("RANGE", [ false, "Enumerate Recently mapped shares"])
-      ]
-    )
+    register_options([
+      OptAddress.new('ADDRESS', [ false, 'IP address to resolve']),
+      OptAddressRange.new('RANGE', [ false, 'IP address range to resolve'])
+    ])
   end
 
   def resolve_ip(ip)
-    ip_ino = Rex::Socket.addr_aton(ip)
-    begin
-      ptr2dns = session.railgun.ws2_32.gethostbyaddr(ip_ino, 4, 2)
+    return unless Rex::Socket.dotted_ip?(ip)
+
+    case session.type
+    when 'powershell'
+      host = cmd_exec("[System.Net.Dns]::GetHostEntry('#{ip}').HostName").to_s
+
+      if host.blank?
+        print_error("Failed to resolve #{ip}")
+        return
+      end
+
+      return host
+    when 'meterpreter'
+      ip_ino = Rex::Socket.addr_aton(ip)
+      ptr2dns = client.railgun.ws2_32.gethostbyaddr(ip_ino, 4, 2)
```

Comment:
It might be worthwhile trying to update this to `GetNameInfo[A|W]`
`gethostbyaddr` is being deprecated, and `GetNameInfo` is much cleaner function to use, it supports unicode names, and in theory, it is supported back to Winxp SP2: 
> The Unicode version of this function available on Windows XP with Service Pack 2 (SP2) and later is [GetNameInfoW](https://learn.microsoft.com/en-us/windows/desktop/api/ws2tcpip/nf-ws2tcpip-getnameinfow).
I'm also worried a bit about the sloppy memread- the return value is never checked and we blindly read 255 bytes.

Reply:
> It might be worthwhile trying to update this to `GetNameInfo[A|W]`

Agreed, but I'll leave this as a task for someone else.

> I'm also worried a bit about the sloppy memread- the return value is never checked and we blindly read 255 bytes.

An existing issue in this module.

I added a check for the return value (for comparison, the `reverse_lookup` module checks the return value and only reads 100 bytes) and now the module works on Meterpreter sessions. I'm not sure why. As that makes no sense, I'm going to assume there was something wrong with my initial testing. I blame DNS as per tradition.

> I noticed this also fails on psexec sessions

Instantly kills psexec Meterpreter sessions for me too.